theano/tensor/signal/pool.py

```diff
@@ -113,30 +114,103 @@ def pool_2d(input, ds, ignore_border=None, st=None, padding=(0, 0),
     return tensor.reshape(output, outshp, ndim=input.ndim)
 
 
-class Pool(OpenMPOp):
+def pool_3d(input, ds, ignore_border=None, st=None, padding=(0, 0, 0),
+            mode='max'):
+    """Downscale the input by a specified factor
+
+    Takes as input a N-D tensor, where N >= 3. It downscales the input image by
+    the specified factor, by keeping only the maximum value of non-overlapping
+    patches of size (ds[0],ds[1],ds[2])
+
+    Parameters
+    ----------
+    input : N-D theano tensor of input images
+        Input images. Max pooling will be done over the 3 last dimensions.
+    ds : tuple of length 3
+        Factor by which to downscale (vertical ds, horizontal ds, depth ds).
+        (2,2,2) will halve the image in each dimension.
+    ignore_border : bool (default None, will print a warning and set to False)
+        When True, (5,5,5) input with ds=(2,2,2) will generate a (2,2,2) output.
+        (3,3,3) otherwise.
+    st : tuple of three ints
+        Stride size, which is the number of shifts over rows/cols/slices to get
+        the next pool region. If st is None, it is considered equal to ds
+        (no overlap on pooling regions).
+    padding : tuple of two ints
+        (pad_h, pad_w, pad_d), pad zeros to extend beyond six borders of the
+        images, pad_h is the size of the top and bottom margins,
+        pad_w is the size of the left and right margins, and pad_d is the size
+        of the depth margins
+    mode : {'max', 'sum', 'average_inc_pad', 'average_exc_pad'}
+        Operation executed on each window. `max` and `sum` always exclude
+        the padding in the computation. `average` gives you the choice to
+        include or exclude it.
+
     """
-    For N-dimensional tensors, consider that the last two dimensions span
-    images. This Op downsamples these images by taking the max, sum or average
-    over different patch.
+    if input.ndim < 3:
+        raise NotImplementedError('pool_3d requires a dimension >= 3')
+    if ignore_border is None:
+        warnings.warn(
+            "pool_3d() will have the parameter ignore_border"
+            " default value changed to True (currently"
+            " False). To have consistent behavior with all Theano"
+            " version, explicitly add the parameter ignore_border=True."
+            " On the GPU, using ignore_border=True is needed to use cuDNN."
+            " When using ignore_border=False and not using cuDNN, the only"
+            " GPU combination supported is when"
+            " `ds == st and padding == (0, 0, 0) and mode == 'max'`."
+            " Otherwise, the convolution will be executed on CPU.",
+            stacklevel=2)
+        ignore_border = False
+    if input.ndim == 5:
```

Comment:
If the op support doing in the Op the extra dimensions, we could here always add a Pool op, but we will need to move that logic that add the reshapes in the optimization that move the op on the GPU to keep them working in that case. Otherwise, the op won't do the extra dimensions itself, it will be done in the graph.